include/onnxruntime/core/framework/tensor.h

```diff
@@ -216,6 +220,10 @@ class Tensor final {
     return buffer_deleter_ != nullptr;
   }
 
+  AllocatorPtr StealDeleter() {
+    return std::move(buffer_deleter_);
```

Comment:
I try to support case when InferenceSession is out of scope but output numpys are still used. if OrtValues are owners of memory they will release it but numpy arrays will still have pointers to the memory. In my understanding a numpy array is final owner of this memory.

Reply:
The OrtValues/Tensors that ORT outputs are allocated from the internal allocators. Those allocators are owned by the InferenceSession. Thus, you typically will have to copy memory to be owned by numpy arrays. That would defeat the purpose of the PR.
It is possible to pre-create output OrtValues on top of the user owned buffers and specified them prior running the code, as is done for input, but for that you would need to always know the shapes.

Reply:
Hello @yuslepukhin!
> It is possible to pre-create output OrtValues on top of the user owned buffers and specify them as args to Run(), as is done for input, but for that you would need to always know the shapes.
Yes, it is other approach described [in the issue](https://github.com/microsoft/onnxruntime/issues/11099). But it requires to follow OrtValue types somehow due to as result it can be not only Tensor but also string inside Tensor, SparseTensor, NonTensor or nothing. We **potentially** can allocated memory inside numpy array only for Tensor. Other problem appears for the approach is undefined shapes (e.g. input shape [-1 3 640 480] in onnx model). It is solved by shape inference mechanism for given input shapes, but nobody described me the mechanism in the issue. I think current approach implemented in the PR is more elegant and does not require these things. More other you suggested it.
> Those allocators are owned by the InferenceSession.
It is not so and in fact this PR solves the problem that you described. I steal the allocator from OrtValue/Tensor and it means that it will not be launched from OrtValue destructor and the memory will not be deallocated. Due to the allocators are kept by shared ptrs they do not die when InferenceSession is realesed. For the latter I create copy of shared ptr (counter is incremented) and keep it by py::capsule mechanism. Thus when numpy array is destructed the allocator is launched, deallocates the memory and realesed by py::capsule (shared ptr counter is decremented).

Reply:
I have not looked at the most recent commits. However, Tensor class is very fundamental for the implementation. I am not in favor of changing it to achieve a narrow API optimization. ORT Allocators are tied to execution providers, EPs in turn are implemented in shared objects and DLLs which will support loading and unloading in the next release.  Furthermore, the Tensor in output may not have an allocator altogether if it is allocated on top of memory is a part of a bigger allocation chunk. The Tensor also many not have an allocator if the user pre-allocated the output on top of the user allocated buffer.
It can also have a custom allocator that was supplied but the user. I am not seeing all these cases well addressed. There are, I am sure, further considerations when changing something like Tensor.
SparseTensor python API illustrates some approaches taken to expose buffers w/o copying data based on existing mechanism. It also takes care of python recounting to make sure that the parent SparseTensor object is not garbage collected before the numpy array being returned to the user. This approach works for both user allocated buffers and buffers from the internal ORT allocators.

Reply:
Hello @yuslepukhin! As I understand the feature is uncompatible with ORT design in current state. I expected that something like this might happen. It looks like that other approaches (memory allocation by numpy arrays before inference or SparseTensor like approach) can have the same problems. I need more time to analyze all things. I suggested to transfer discussion of universal feature and approaches back to [the issue](https://github.com/microsoft/onnxruntime/issues/11099).
If we can not use universal approach now, could we hide the feature by some condition as specific thing as first step? I see three approaches for this: a. check EP type (for me it is TVM EP), b. create new session option and check it, c. use #ifdef for new experimental macros defined in cmake. Which one is the best for specific approach?

Reply:
Hello @yuslepukhin! Could you response on my comment?

Reply:
I think you can reasonably implement the feature and derive the expected benefit, provided you make sure that inference session object remains valid. Otherwise, create output values on top of the numpy memory in which case you would have no dependancy. There is not need to hide anything.